SARAALERT-1192: Performance testing

jkufro opened the pull request:
# Description
Jira Ticket: SARAALERT-1192

This PR shows an example JMeter test automated in CI, as well as a standalone example benchmark script for the send assessments job.

Rails 6.1 has this [benchmark generator built-in](https://github.com/rails/rails/tree/master/railties/lib/rails/generators/rails/benchmark) that uses the [benchmark-ips](https://github.com/evanphx/benchmark-ips) gem. We are also free to modify the file to use the vanilla benchmark library (https://ruby-doc.org/stdlib-2.5.0/libdoc/benchmark/rdoc/Benchmark.html)

IMO JMeter is a better solution for load and stress testing, while the rails benchmarking example does well for understanding performance of highly targeted slices of the code with detailed CPU & RAM logging with stackprof and memory_profiler. We would likely benefit by leveraging both methods of testing our app. If we want, then we could also run stackprof with rack so that when we run JMeter we can capture data there as well.

The JMeter testing has been setup to work automatically through our mirrored GitLab repo on all branches. If the database does not need to be rebuilt, then the test takes about 5 minutes. If the database does need to be rebuilt because of a migration failure, then the population of the 1M monitoree database takes the total test time up to about 30 minutes. I set it up to use the 1M monitoree database so that we can be reasonably close to production loads in our JMeter testing. One specific place that might need some discussion/tweaking is the number of threads and loops for the JMeter test.

Ruby JMeter (write JMeter test plans in ruby to run or export): https://github.com/flood-io/ruby-jmeter
JMeter Site: https://jmeter.apache.org/

__JMeter Setup (MacOS)__:
```bash
# there is an issue with the latest, so we need to point to a previous version
brew edit jmeter

# change these lines 
...
url "https://www.apache.org/dyn/closer.lua?path=jmeter/binaries/apache-jmeter-5.2.1.tgz"
mirror "https://archive.apache.org/dist/jmeter/binaries/apache-jmeter-5.2.1.tgz"
sha256 "bbb3cb5fe0b16c1fa139727063f18d9aca6890c7edf53f3a614b8f929f1f1de9"
...

# install jmeter
brew install jmeter

# added to ~.zprofile for covenience
alias jmeter='open /usr/local/bin/jmeter'

```

## Installed JMeter Plugins
On Mac:`Options > Plugins Manager`

<img width="248" alt="Screen Shot 2021-01-20 at 9 37 19 AM" src="https://user-images.githubusercontent.com/24628687/105206014-188f3500-5b03-11eb-9f52-cbe70b01f33b.png">

## JMeter Metrics
The added plugins allow us to generate the below charts in JMeter.

### Response Latencies Over Time
<img width="875" alt="Screen Shot 2021-01-20 at 10 19 35 AM" src="https://user-images.githubusercontent.com/24628687/105211005-ff898280-5b08-11eb-9175-46c64ef73d0b.png">

### Response Codes per Second
<img width="878" alt="Screen Shot 2021-01-20 at 10 19 43 AM" src="https://user-images.githubusercontent.com/24628687/105211021-057f6380-5b09-11eb-89d4-950518ef3c7d.png">

### Response Times Distribution
<img width="876" alt="Screen Shot 2021-01-20 at 10 19 52 AM" src="https://user-images.githubusercontent.com/24628687/105211041-0adcae00-5b09-11eb-8f6a-40a817999bb3.png">

### Response Times Over Time
<img width="880" alt="Screen Shot 2021-01-20 at 10 20 01 AM" src="https://user-images.githubusercontent.com/24628687/105211053-0fa16200-5b09-11eb-912e-0b92ddc67f13.png">

### Response Times Percentiles
<img width="880" alt="Screen Shot 2021-01-20 at 10 20 12 AM" src="https://user-images.githubusercontent.com/24628687/105211071-162fd980-5b09-11eb-87a5-83a1b926f9d1.png">

### Transactions per Second
<img width="877" alt="Screen Shot 2021-01-20 at 10 20 20 AM" src="https://user-images.githubusercontent.com/24628687/105211098-1af48d80-5b09-11eb-9577-1314fd36b849.png">


# Important Changes
Please list important files (meaning substantial or integral to the PR) along with a list of the general changes that should be highlighted for reviewers.

`example_file.js`
- Example change (ex: refactored import function)

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):
* [ ] Chrome
* [ ] Firefox
* [ ] Safari
* [ ] IE11

Please describe the tests needed to verify your changes. Provide instructions for reproducing these tests. List any relevant details for your test configuration.

Bialogs reviewed (changes requested):
I would like to see this get run against the production configuration of the system, not just a `rails s` in development mode.

jkufro commented on script/benchmarks/send_assessments_job_benchmark.rb:
```diff
@@ -0,0 +1,45 @@
+$stdout = File.new(benchmark_file, 'w')
+$stdout.sync = true
+
+Benchmark.bm(20) do |x|
```
@Bialogs I found in my testing that when benchmarking ips and stackprof are used together that stackprof breaks and becomes unusable. Since I figured stackprof is preferred over IPS I am using vanilla benchmarking here

Bialogs replied:
That's fine. I think the flamegraph is nice to have. Any thoughts on https://ruby-doc.org/stdlib-2.7.1/libdoc/benchmark/rdoc/Benchmark.html#method-c-bmbm

jkufro replied:
Since this test in particular is expected to take anywhere from 30-60 minutes (or slightly more) then running the test twice with `bmbm` would result in double the testing time.

Bialogs replied:
Yeah great point...With the amount of action happening in the `perform_now` it should be fine. Maybe `bmbm` is more for a microbenchmark scenario.

holmesie commented on script/jmeter/01_public_health_dashboard_jmx_test.rb:
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require 'ruby-jmeter'
```
Thoughts for keeping these `scripts` separate vs. as rake tasks?

jkufro replied:
The`scripts` folder gets generated when your run `rails g benchmark ...` and I kept that for the both the benchmarking folder and jmeter folder. I wouldn't really mind if we made them rake tasks, but when we want to generate new benchmarks we would just need to manually move them I think.

holmesie replied:
Ah I see. I think it's fine as is. Thank you for answering my question!

holmesie reviewed (changes requested):
This looks excellent, and I'd like to incorporate this as soon as possible. Just one question I left inline.

Bialogs reviewed (changes requested):
You need to look at the `public_health_controller#patients` route...

```
Started POST "/public_health/patients" for 127.0.0.1 at 2021-02-01 13:45:26 -0800
Processing by PublicHealthController#patients as HTML
  Parameters: {"query"=>{"workflow"=>"exposure", "tab"=>"all", "jurisdiction"=>1, "scope"=>"all", "user"=>nil, "search"=>"ryan", "page"=>"[FILTERED]", "entries"=>25, "tz_offset"=>480}, "cancelToken"=>{"promise"=>{}}, "public_health"=>{"query"=>{"workflow"=>"exposure", "tab"=>"all", "jurisdiction"=>1, "scope"=>"all", "user"=>nil, "search"=>"ryan", "page"=>"[FILTERED]", "entries"=>25, "tz_offset"=>480}, "cancelToken"=>{"promise"=>{}}}}
```

jkufro commented:
> You need to look at the `public_health_controller#patients` route...
> 
> ```
> Started POST "/public_health/patients" for 127.0.0.1 at 2021-02-01 13:45:26 -0800
> Processing by PublicHealthController#patients as HTML
>   Parameters: {"query"=>{"workflow"=>"exposure", "tab"=>"all", "jurisdiction"=>1, "scope"=>"all", "user"=>nil, "search"=>"ryan", "page"=>"[FILTERED]", "entries"=>25, "tz_offset"=>480}, "cancelToken"=>{"promise"=>{}}, "public_health"=>{"query"=>{"workflow"=>"exposure", "tab"=>"all", "jurisdiction"=>1, "scope"=>"all", "user"=>nil, "search"=>"ryan", "page"=>"[FILTERED]", "entries"=>25, "tz_offset"=>480}, "cancelToken"=>{"promise"=>{}}}}
> ```

@Bialogs see latest commits for fix

jkufro commented on script/assert_jmeter_results.rb:
```diff
@@ -0,0 +1,148 @@
+  {
+    label: '01_03_SA_dashboard',
+    failure_thresholds: {
+      avg_latency: 4_000,
```
That's how I did it

kylietmo replied:
Works for me - we can adjust as needed as we get ramped up on these.

jkufro commented on script/jmeter/01_public_health_dashboard_jmx_test.rb:
```diff
@@ -0,0 +1,134 @@
+  end
+
+  # end.run(path: '/usr/local/bin/', gui: false)
+  # end.jmx(file: 'tmp/public_health_dashboard_jmx_test.jmx')
```
I left in for ease on running the tests locally

kylietmo replied:
Good enough for me.

jkufro commented:
Currently investigating an issue with CI testing of the send assessments job. JMeter parts of PR should be ready for review

jkufro commented on Gemfile:
```diff
@@ -107,13 +107,17 @@ gem 'audited'
 gem 'order_as_specified'
 
 group :development, :test do
+  gem 'benchmark-ips'
```
ips is added automatically when using rails benchmark generators - `rails g benchmark foo`

Bialogs replied:
Good enough reason to keep it around. Generators are nice. What about where you `require` in the actual Benchmarking script? Is it needed there?

jkufro replied:
Just removed from script

kylietmo commented on app/jobs/send_assessments_job.rb:
```diff
@@ -8,11 +8,14 @@ def perform(*_args)
       next
+    ensure
+      # Trigger garbage collect every 2,000 patients
+      GC.start if (indx % 2_000).zero?
```
What was the reason for this change in this PR?

jkufro replied:
On the GitLab runner that is doing the job I found that ruby was not garbage collecting fast enough, eats up all the memory, and then gets killed by the system before it can finish. This kicks off garbage collection every 2,000 patients to avoid this issue. 

As more support for this, I'm running the job on my local machine with the 1M database and it does the same without GC being invoked manually (currently @ 25GB and rising, while with GC it hovers somewhere between 0.5 and 1GB).
<img width="964" alt="Screen Shot 2021-02-18 at 8 06 54 AM" src="https://user-images.githubusercontent.com/24628687/108376467-44641000-71c0-11eb-9745-77ad28f9cf7d.png">

Bialogs commented on script/benchmark.rb:
```diff
@@ -0,0 +1,70 @@
+
+  benchmark_report = nil
+
+  ActiveRecord::Base.transaction do
```
Future versions of Rails will have some performance optimizations based on if queries are run in transactions. Is there any way that we can remove the transaction and get the database back into its original state between benchmarks instead of a transaction?

jkufro replied:
See latest changes that address this

Bialogs commented on script/run_jmeter_tests.sh:
```diff
@@ -0,0 +1,19 @@
+    kill -9 $(cat tmp/pids/server.pid)
+fi
+
+ruby script/assert_jmeter_results.rb
```
Newline at EOF :)

jkufro commented:
@kylietmo @Bialogs @holmesie I made it so that memory_profiler could be disabled via an environment variable in the benchmark.rb script. Also added a warning that gets printed to the console about memory_profiler memory issues if the number of patients is > 100,000. Memory_profiler is now disabled for the GitLab CI benchmarks. The most recent pipeline can be found [here](https://gitlab.mitre.org/2019-ncov/disease-trakker/-/pipelines/287825). 

I also removed the GC.start statements that were previously added in.
